src/compiler/emitter.ts

```diff
@@ -2091,16 +2106,54 @@ namespace ts {
             return statements.length;
         }
 
+        function emitPrologueDirectivesIfNeeded(sourceFileOrBundle: Bundle | SourceFile) {
+            if (sourceFileOrBundle.kind === SyntaxKind.SourceFile) {
+                setSourceFile(sourceFileOrBundle as SourceFile);
+                emitPrologueDirectives((sourceFileOrBundle as SourceFile).statements);
+            }
+            else {
+                const seenProgolueDirectives = createMap<String>();
+                for (const sourceFile of (sourceFileOrBundle as Bundle).sourceFiles) {
+                    setSourceFile(sourceFile);
+                    emitPrologueDirectives(sourceFile.statements, /*startWithNewLine*/ true, seenProgolueDirectives);
+                }
+            }
+        }
+
+        function emitShebangIfNeeded(sourceFileOrBundle: Bundle | SourceFile) {
+            if (sourceFileOrBundle.kind === SyntaxKind.SourceFile) {
+                emitShebangInSourceFile(sourceFileOrBundle as SourceFile);
+            }
+            else {
+                for (const sourceFile of (sourceFileOrBundle as Bundle).sourceFiles) {
+                    // Emit only the first encountered shebang
+                    if (emitShebangInSourceFile(sourceFile)) {
+                        break;
+                    }
+                }
+            }
+
+            function emitShebangInSourceFile(sourceFile: SourceFile): boolean {
+                const shebang = getShebang(sourceFile.text);
+                if (shebang) {
+                    write(shebang);
+                    writeLine();
+                    return true;
+                }
+            }
+        }
+
         //
         // Helpers
         //
 
-        function emitShebang() {
-            const shebang = getShebang(currentSourceFile.text);
-            if (shebang) {
-                write(shebang);
-                writeLine();
+        function getIndexOfFirstNonPrologueDirectives(statements: Statement[]): number {

```

Comment:
Could we generalize this using `findIndex`? For example:

```ts
const index = findIndex(statements, statement => !isPrologueDirective(statement))
return index === -1 ? statements.length : index;
```

Alternatively we can add a parameter to `countWhere` to stop at the first predicate failure:

```ts
countWhere(statements, isPrologueDirective, /*initialSpan*/ true);

// core.ts
export function countWhere<T>(array: T[], predicate: (x: T, i: number) => boolean, initialSpan?: boolean): number {
    let count = 0;
    if (array) {
        for (let i = 0; i < array.length; i++) {
            const v = array[i];
            if (predicate(v, i)) {
                count++;
            }
            else if (initialSpan) {
                break;
            }
        }
    }
    return count;
}
```